### ISSUE TITLE
sct_image - stitch - cuts off images in axial/sagittal planes

### ISSUE THREAD

**Issue by jqmcginnis** [spinalcordtoolbox/spinalcordtoolbox#3983]
### Description

Hey team :slightly_smiling_face: 
When using our new feature `sct_image -stitch` on a large clinical cohort, I realized that the axial/sagittal views are cut off for scans with different fields of view. Often this might be subtle, e.g. if you focus on the spinal cord / central image part when stitching images, it might not be obvious. However, for patients with sclerosis, it is highly visible as parts of the original image are cut off.

### Steps to Reproduce

0. Install SCT v5.8 Spinal Cord Toolbox (git-master-a0aff1115d36dcae447360e3b66811f4b7105bee)

1. Run sct_image -stitch, e.g.:

`sct_image -i sub-m123456_ses-19990101_acq-ax_chunk-1_T2w.nii.gz sub-m123456_ses-19990101_acq-ax_chunk-2_T2w.nii.gz sub-m123456_ses-19990101_acq-ax_chunk-3_T2w.nii.gz -o test.nii.gz -stitch`

This will output an image with visible cutoff (c.f. left part).

**Expected behavior:** 
Produce a stitched image with intact axial / sagittal planes, i.e. covering the complete space of the input images.
![1](https://user-images.githubusercontent.com/33037028/208702426-192c7900-8539-4af2-aaf3-a62ed0f86b3f.png)

**Actual behavior:** <What actually happens>
Produces a stitched image with cut-off axial / sagittal planes, i.e. only partially covering the space of the input images.
 ![2](https://user-images.githubusercontent.com/33037028/208702171-c25f3d22-8cba-450d-9ba6-0cfaf2a8af15.png).

## Fix

Generously padding the image prior to using `sct_image -stitch` does the trick for now:

```
sct_image -i sub-m123456_ses-19990101_acq-ax_chunk-1_T2w.nii.gz -pad 400,400,0 -o sub-m123456_ses-19990101_acq-ax_chunk-1_T2w_pad.nii.gz
sct_image -i sub-m123456_ses-19990101_acq-ax_chunk-2_T2w.nii.gz -pad 400,400,0 -o sub-m123456_ses-19990101_acq-ax_chunk-2_T2w_pad.nii.gz
sct_image -i sub-m123456_ses-19990101_acq-ax_chunk-3_T2w.nii.gz -pad 400,400,0 -o sub-m123456_ses-19990101_acq-ax_chunk-3_T2w_pad.nii.gz
```
However, there is also a more sound approach to padding the images. If I debug the original stitching module (or add additional printouts), it is quite reasonable why the images are cropped/cutoff in the stitching module: Glocker et al. only pad in z-dimension.
But the leftmost and rightmost coordinates are different for x/y as well:

```
Image 1:
...
ymin1=66.0269
ymax1=186.027
...

Image 2:
...
ymin2=100.538
ymax2=220.538
...

```

Thus, a valid calculation for e.g. y-padding would be `n_voxels=(Y_max-Y_min)/spacing = (220,538 − 66,0269) / 0.208033 = 743 voxels`. If the image spans 640 voxels, we can determine the asymmetric padding as well, e.g. `y_pad_left_img2= (ymin2-ymin1)/spacing = 166`, and use asymmetric padding` sct_image -pad-asym` - to mitigate the shortcomings of the stitching tool.

### Proposed solution:
Add asyymetric  padding directly in the `sct_image -stitch` so that the final user does not have to apply it!

Let me know what you think :slightly_smiling_face: 
Julian

**Comment by joshuacwnewton** [comment]
Thank you so much for the detailed investigation + proposed solution, @jqmcginnis!

> If I debug the original stitching module (or add additional printouts), it is quite reasonable why the images are cropped/cutoff in the stitching module: Glocker et al. only pad in z-dimension.

Aha! So, if I'm understanding correctly, this flaw is inherent to the `stitching` binary itself (and thus would be present even when running `stitching` outside of SCT?)

> Thus, a valid calculation for e.g. y-padding would be `n_voxels=(Y_max-Y_min)/spacing = (220,538 − 66,0269) / 0.208033 = 743 voxels`. If the image spans 640 voxels, we can determine the asymmetric padding as well, e.g. `y_pad_left_img2= (ymin2-ymin1)/spacing = 166`, and use asymmetric padding` sct_image -pad-asym` - to mitigate the shortcomings of the stitching tool.

I must admit that I don't fully understand the proposed solution, though. How were the `y_min` and `y_max` values determined, and what coordinate system do they represent?

**Comment by jqmcginnis** [comment]
>Aha! So, if I'm understanding correctly, this flaw is inherent to the stitching binary itself (and thus would be present even when running stitching outside of SCT?)

Yes, indeed! :slightly_smiling_face:  The original paper is intended for whole body MR images - the original protocol covers the whole body by scanning in axial direction, and it uses the same Field of View / scanner space for the whole body as it covers it entirely. This is different from the sequence/protocol the neuroradiologists use in Munich for our spine cohort. Here, the MR technician optimizes the Field of View for each stack (cervical, lumbar and thoracic) as the FOV is limited, and its size is fixed. As the spine is curved, we end up imaging (slightly) different x-/y-coordinate planes in each of the stacks.  

>How were the y_min and y_max values determined, and what coordinate system do they represent?

Sorry, for forgetting to include this. I debugged these coordinates directly in stitching by including the following code in the stitching sources and recompiling/rebuilding the binary.
```
  // determine the physical size of all images!

  for (int i = 0; i < files.size(); i++)
  {
    auto image= itkio::load(files[i]);

    auto min_extent_x = image.origin()[0];
    auto max_extent_x = image.origin()[0]+image.sizeX() * image.spacing()[0];
    auto min_extent_y = image.origin()[1];
    auto max_extent_y = image.origin()[1]+image.sizeY() * image.spacing()[1];
    auto min_extent_z = image.origin()[2];
    auto max_extent_z = image.origin()[2]+image.sizeZ() * image.spacing()[2];

  }
```
This code snippet basically calculates the scanner space coordinates (which are used by the stitching binary as it operates in the scanner space, not the voxel space). The z-coordinates are also used for padding of the images in z-direction, however stitching lacks x-/y-padding. 

If we want to mitigate this without changing stitching's code base, we could use the following steps for all images:

1. Read the NIFTI Headers of all images in Image Class (I checked it includes the hdr)
2. Use origin and spacing to calculate x_min, x_max, y_min, y_max in scanner coordinates for each image
3. Make sure to pad all images in x/y-direction so that each of the images covers the complete voxel space (min/max scanner coordinates) - this should translate into asymmetric padding of the images. 

Does this answer your question?

**Comment by joshuacwnewton** [comment]
> This is different from the sequence/protocol the neuroradiologists use in Munich for our spine cohort.

Wonderful! Thank you so much for the added context. It's fascinating learning about what your group does. :smile: 

> Does this answer your question?

Absolutely! Thank you for providing the precise equations that you used to debug this problem. :hearts:

----

One follow-up question: How urgent is this fix? In other words, how much of an inconvenience is it to use the manual padding workaround that you've come up with?

(If it's a hindrance to your productivity/results, I'll happily prioritize a fix for this highly!)

**Comment by jqmcginnis** [comment]
>One follow-up question: How urgent is this fix? In other words, how much of an inconvenience is it to use the manual padding workaround that you've come up with?

Since it should not be very complicated to implement a bugfix and I am familiar with the problem and setup, I am happy to work on this in the next few days :slightly_smiling_face:.

**Comment by joshuacwnewton** [comment]
That would be lovely! Thank you so much for offering. :hearts:

**Comment by joshuacwnewton** [comment]
Closing as a duplicate of #5107 (oops, we repeated ourselves in that issue)
